docs/testing_using_Chai_&_Mocha.md

```diff
@@ -0,0 +1,161 @@
+  it("Sample tests with mocha and chai", async function () {
+    var foo = 'bar'
+    var beverages = { tea: [ 'chai', 'matcha', 'oolong' ] };
+    expect(foo).to.be.a('string');
+    expect(foo).to.equal('bar');
+    expect(foo).to.have.lengthOf(3);
+    expect(beverages).to.have.property('tea').with.lengthOf(3);
+  });
+});
+```
+
+## Run tests
+
+Once done with writing the tests, right click on file name in `File Explorers` plugin. It will show some options along with option to `Run with Mocha`.
+
+![](images/run_with_mocha_option.png)
+
+Click on `Run with Mocha` option, tests will be executed and result will be shown on Terminal.
+
+![](images/run_with_mocha_result.png)
+
+## Test a contract
+
+Similarly unit tests can be written to test the functionality of a smart contract. An example to test default `1_Storage.sol` contract can be as:
+
+```
+const { expect } = require("chai");
+
+describe("Storage", function () {
+  it("test initial value", async function () {
+    const metadata = JSON.parse(await remix.call('fileManager', 'getFile', 'contracts/artifacts/Storage.json'))
+    const signer = (new ethers.providers.Web3Provider(web3Provider)).getSigner()
+    let Storage = new ethers.ContractFactory(metadata.abi, metadata.data.bytecode.object, signer);
+    let storage = await Storage.deploy();
+    console.log('storage contract Address: ' + storage.address);
+    await storage.deployed()
+    expect((await storage.retrieve()).toNumber()).to.equal(0);
+  });
+
+  it("test updating and retrieving updated value", async function () {
+    const metadata = JSON.parse(await remix.call('fileManager', 'getFile', 'contracts/artifacts/Storage.json'))
+    const signer = (new ethers.providers.Web3Provider(web3Provider)).getSigner()
+    let Storage = new ethers.ContractFactory(metadata.abi, metadata.data.bytecode.object, signer);
+    let storage = await Storage.deploy();
+    await storage.deployed()
+    const setValue = await storage.store(56);
+    await setValue.wait();
+    expect((await storage.retrieve()).toNumber()).to.equal(56);
+  });
+
+  it("fail test updating and retrieving updated value", async function () {
+    const metadata = JSON.parse(await remix.call('fileManager', 'getFile', 'contracts/artifacts/Storage.json'))
+    const signer = (new ethers.providers.Web3Provider(web3Provider)).getSigner()
+    let Storage = new ethers.ContractFactory(metadata.abi, metadata.data.bytecode.object, signer);
+    let storage = await Storage.deploy();
+    await storage.deployed()
+    const setValue = await storage.store(56);
+    await setValue.wait();
+    expect((await storage.retrieve()).toNumber()).to.equal(55);
+  });
+});
+```
+
+Result will be as:
+
+![](images/run_with_mocha_storage_test.png)
+
+## Debugging a test transaction
+
+To debug a transaction in one of the tests can be done by printing the tranasation hash and using the [Remix Debugger plugin](https://remix-ide.readthedocs.io/en/latest/tutorial_debug.html).
```

Comment:
Some small edits:
To debug a transaction in one of the tests,  print the transaction hash and input that in the [Remix Debugger plugin](https://remix-ide.readthedocs.io/en/latest/tutorial_debug.html).

Reply:
updated